pkg/images/enricher/enricher_impl.go

```diff
@@ -711,11 +715,13 @@ func filterRegistriesBySource(requestSource *RequestSource, registries []registr
 }
 
 func getMatchingRegistries(registries []registryTypes.ImageRegistry,
-	image *storage.Image) ([]registryTypes.ImageRegistry, error) {
-	var matchingRegistries []registryTypes.ImageRegistry
+	image *storage.Image) (map[string][]registryTypes.ImageRegistry, error) {
+	matchingRegistries := make(map[string][]registryTypes.ImageRegistry, len(image.GetNames().GetNames()))
 	for _, registry := range registries {
-		if registry.Match(image.GetName()) {
-			matchingRegistries = append(matchingRegistries, registry)
+		for _, name := range image.GetNames().GetNames() {
+			if registry.Match(image.GetName()) {
+				matchingRegistries[name.GetFullName()] = append(matchingRegistries[name.GetFullName()], registry)
```

Comment:
This seems disadvantageous. If you have multiple images referring to the same registry, you will return the registry multiple times, and make multiple calls to the same registry for different names. This seems to be unavoidable from the standpoint of the cosign library (you can only fetch signatures for a single reference), but I guess it could be argued from a retry budget standpoint that retry budget for transient errors should be per registry, not per reference.

~Therefore, grouping by registry seems to make more sense to me.~ I wrote this before my above comments that registries should be sorted. Not sure what would be the right approach here. However, it seems unnecessarily complex to return a `map[string][]Registry`. Couldn't you just have left this function as-is except for replacing the `image` parameter with a `name` parameter, and then just call it in every iteration of the outer loop in l.608?

Reply:
I agree that it's unfortunate regarding having to make multiple calls to the same registry for the different image names.

I've change it up a bit to follow your suggestion with calling getMatchingRegistries within the for-loop, with minor caveats:

before trying to do anything, we first validate whether at least one registry exists (irrespective of the image name) with a new helper - which will error out with the same conditions as previously for getMatchingRegistries.
within the for-loop, iff for one particular image name no matching integrations are given, we are going to log it and skip fetching signatures for that particular image name - I don't think this mandates that we error out completely of the image enrichment process. Say, we would have the case where an image is having the same digest in a OpenShift internal registry and a public registry, we would always error out enriching that particular image in the central context.